src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80CompoundDirectory.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.KNN80Codec;
+
+import lombok.Getter;
+import org.apache.lucene.codecs.CompoundDirectory;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+
+import java.io.IOException;
+import java.util.Set;
+
+public class KNN80CompoundDirectory extends CompoundDirectory {
+
+    @Getter
+    private CompoundDirectory delegate;
+    @Getter
+    private Directory dir;
+
+    public KNN80CompoundDirectory(CompoundDirectory delegate, Directory dir) {
+        this.delegate = delegate;
+        this.dir = dir;
+    }
+
+    @Override
+    public void checkIntegrity() throws IOException {
+        delegate.checkIntegrity();
+    }
+
+    @Override
+    public String[] listAll() throws IOException {
+        return delegate.listAll();
+    }
+
+    @Override
+    public long fileLength(String name) throws IOException {
+        return delegate.fileLength(name);
+    }
+
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        return delegate.openInput(name, context);
```

Comment:
For future extendability, could we check if this is a k-NN file and open it from the directory? I know we typically dont open k-NN files via openInput, but if we do, this might fail:
```
        if (KNNEngine.getEnginesThatCreateCustomSegmentFiles().stream().anyMatch(engine -> name.endsWith(engine.getCompoundExtension()))) {
            return directory.openInput(name, context);
        }

        return delegate.openInput(name, context);
```

Reply:
@jmazanec15 why we want to do this? this can lead to unnecessary mapping of the file in memory. My thoughts will be not to do this.

Reply:
I dont think anyone should call openInput on k-NN files. But if they do, `return delegate.openInput(name, context);
` will throw exception because file isnt actually in compound file

Reply:
I am not sure if exception will come. @luyuncheng have you seen an exception here? I think it will open the input. 

But even if there is no exception we should not open the input, it is just unnecessary mapping of file. One we start to use IndexInput to read the graph file(ref: https://github.com/opensearch-project/k-NN/issues/1951) we can start opening the files here.

Reply:
i did not find any exception. but i do want to implement an nativeEngine indexInput like #1951 says. 
how about:
1. we return `return delegate.openInput(name, context);` 
2. and added a assert when name is native engines. 
3. also added an TODO comments here: TODO: using the native engine IndexInput.

@navneet1v @jmazanec15

Reply:
@navneet1v See how we write: https://github.com/luyuncheng/k-NN-1/blob/DocValuesProducers/src/main/java/org/opensearch/knn/index/codec/KNN80Codec/KNN80CompoundFormat.java#L48-L53. We dont use the delegate to write the file, we use directory directly.

Reply:
@jmazanec15 I understand we would be directly writing the file. I have tested a small code change where in KNNWeight class I was opening .hnsw file using IndexInput and it worked. I didn't get the exception. So want to know when you say there will be exception what will be that exception.

Neverthless I believe we should not even open the nativeIndex files in the producer as it will unnecessary map the file on RAM.

Reply:
In your test was it an .hnsw file or .hnswc? I think .hnswc should fail with this line